ranger/core/actions.py

```diff
@@ -303,91 +321,128 @@ def get_macros(self):  # pylint: disable=too-many-branches,too-many-statements
             macros['datadir'] = self.fm.datapath()
         macros['space'] = ' '
 
-        if self.fm.thisfile:
-            macros['f'] = self.fm.thisfile.relative_path
-        else:
-            macros['f'] = MACRO_FAIL
+        macros['f'] = lambda: self.fm.thisfile.relative_path if self.fm.thisfile else MACRO_FAIL
 
-        if self.fm.thistab.get_selection:
-            macros['p'] = [os.path.join(self.fm.thisdir.path, fl.relative_path)
-                           for fl in self.fm.thistab.get_selection()]
-            macros['s'] = [fl.relative_path for fl in self.fm.thistab.get_selection()]
-        else:
-            macros['p'] = MACRO_FAIL
-            macros['s'] = MACRO_FAIL
+        def get_path():
+            if self.fm.thistab and self.fm.thistab.get_selection:
+                res = [os.path.join(self.fm.thisdir.path, fl.relative_path)
+                       for fl in self.fm.thistab.get_selection()]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['p'] = get_path
 
-        if self.fm.copy_buffer:
-            macros['c'] = [fl.path for fl in self.fm.copy_buffer]
-        else:
-            macros['c'] = MACRO_FAIL
+        def get_selection():
+            if self.fm.thistab and self.fm.thistab.get_selection:
+                res = [fl.relative_path for fl in self.fm.thistab.get_selection()]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['s'] = get_selection
 
-        if self.fm.thisdir.files:
-            macros['t'] = [fl.relative_path for fl in self.fm.thisdir.files
-                           if fl.realpath in self.fm.tags or []]
-        else:
-            macros['t'] = MACRO_FAIL
+        macros['c'] = lambda: [
+            fl.path for fl in self.fm.copy_buffer] if self.fm.copy_buffer else MACRO_FAIL
 
-        if self.fm.thisdir:
-            macros['d'] = self.fm.thisdir.path
-        else:
-            macros['d'] = '.'
+        def get_tagged():
+            if self.fm.thisdir and self.fm.thisdir.files:
+                res = [fl.relative_path for fl in self.fm.thisdir.files
+                       if fl.realpath in self.fm.tags or []]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['t'] = get_tagged
+
+        macros['d'] = lambda: self.fm.thisdir.path if self.fm.thisdir else '.'
 
         # define d/f/p/s macros for each tab
+        tabdir = None
         for i in range(1, 10):
             try:
                 tab = self.fm.tabs[i]
             except KeyError:
                 continue
+            if not tab:
```

Comment:
How can this occur exactly if there's no `KeyError`?

Reply:
I am too not sure why I added this in the original commit. Macros seem to be ok without it, I just tested so yeah, I'll remove it.